livekit-agents/livekit/agents/voice/ivr/ivr_activity.py

```diff
@@ -0,0 +1,214 @@
+
+    def reset(self) -> None:
+        self._vectorizer = TfidfVectorizer()
+        self._transcribed_chunks = []
+        self._num_consecutive_similar_chunks = 0
+
+    async def aclose(self) -> None:
+        self._session.off("user_input_transcribed", self._on_user_input_transcribed)
+
+    def _on_user_input_transcribed(self, ev: UserInputTranscribedEvent) -> None:
+        if not ev.is_final:
+            return
+
+        self._transcribed_chunks.append(ev.transcript)
+        if len(self._transcribed_chunks) > self._window_size:
+            self._transcribed_chunks = self._transcribed_chunks[-self._window_size :]
+
+        # NOTE: currently this is O(n^2) in the number of chunks, let's figure out a more efficient
+        # way if this become a bottleneck later.
+        doc_matrix = self._vectorizer.fit_transform(self._transcribed_chunks)
+        doc_similarity = cosine_similarity(doc_matrix)
+        last_chunk_similarity = doc_similarity[-1][:-1]
+
+        if np.max(last_chunk_similarity) > self._similarity_threshold:
+            self._num_consecutive_similar_chunks += 1
+        else:
+            self._num_consecutive_similar_chunks = 0
+
+        if self.loop_detected:
+            self.emit("loop_detected", None)
+
+
+class SilenceDetector(EventEmitter[SilenceDetectorEventTypes]):
+    """Silence detector.
+
+    This detector checks for silence in the user / agent interaction.
+
+    Args:
+        session: The agent session.
+        max_silence_duration: The maximum duration of silence to detect in seconds. Default ``5.0`` seconds.
+    """
+
+    def __init__(
+        self,
+        session: AgentSession,
+        *,
+        max_silence_duration: float = 5.0,
+    ) -> None:
+        super().__init__()
+        self._session = session
+        self._max_silence_duration = max_silence_duration
+        self._current_user_state: Optional[str] = None  # noqa: UP007
+        self._current_agent_state: Optional[str] = None  # noqa: UP007
+        self._debounced_emit = Debounced(self._emit_silence_detected, self._max_silence_duration)
+
+    async def start(self) -> None:
+        self._session.on("user_state_changed", self._on_user_state_changed)
+        self._session.on("agent_state_changed", self._on_agent_state_changed)
+
+    async def aclose(self) -> None:
+        self._debounced_emit.cancel()
+        self._session.off("user_state_changed", self._on_user_state_changed)
+        self._session.off("agent_state_changed", self._on_agent_state_changed)
+
+    def _on_user_state_changed(self, ev: UserStateChangedEvent) -> None:
+        self._current_user_state = ev.new_state
+        self._schedule_check()
+
+    def _on_agent_state_changed(self, ev: AgentStateChangedEvent) -> None:
+        self._current_agent_state = ev.new_state
+        self._schedule_check()
+
+    def _schedule_check(self) -> None:
+        if self._current_user_state == self._current_agent_state == "listening":
+            logger.info(
```

Comment:
The logs are quite noisy and not necessarily useful.

We just need a IVR detected log